daliuge-engine/test/test_drop.py

```diff
@@ -747,8 +773,125 @@ def checkDropStates(aStatus, dStatus, eStatus, lastByte):
         checkDropStates(
             DROPStates.COMPLETED, DROPStates.COMPLETED, DROPStates.COMPLETED, b"k"
         )
+        self.assertEqual(b"ejk", droputils.allDropContents(d))
+
+    def test_objectAsNormalAndThreadedAsyncStreamingInput(self):
+        self._test_objectAsNormalAndAsyncStreamingInput(False)
+
+    def test_objectAsNormalAndMPAsyncStreamingInput(self):
+        self._test_objectAsNormalAndAsyncStreamingInput(True)
+
+    def _test_objectAsNormalAndAsyncStreamingInput(self, use_multiprocessing):
+        """
+        A test that checks that a DROP can act as normal and async streaming input
+        of different AppDROPs at the same time. We use the following graph:
+
+        A --|--> B --> D
+            |--> C --> E
+
+        Here B uses A as a streaming input, while C uses it as a normal
+        input
+        """
+
+        class QueueAsyncStream(AsyncIterator):
+            _q = multiprocessing.Queue()
+            _end = multiprocessing.Event()
+            async def __anext__(self):
+                while True:
+                    if self._q.qsize() > 0:
+                        return self._q.get()
+                    elif self._end.is_set():
+                        raise StopAsyncIteration
+                    else:
+                        await asyncio.sleep(0)
+
+            def append(self, data):
+                while not self._q.full():
+                    try:
+                        self._q.put(data, block=True, timeout=1)
+                        break
+                    except queue.Full:
+                        pass
+
+            def end(self):
+                self._end.set()
```

Comment:
Previously I used a deque here however multiprocess safe issues lead me to trial a queue. Here I've also set the stream base instead to AsyncIterator instead of AsyncIterable where the primary difference is the iterator can only be iterated once. This greatly simplifies the lifecycle of around when to pop data and embraces shared ownership of stream data.

This implementation assumes putting in one process and getting in another has minimal overhead.